File: solutions/php/bottle-song/1/BottleSong.php
<?php declare(strict_types=1);

final class BottleSong
{
    private const NUMERALS = [
        0 => "no",
        1 => "one",
        2 => "two",
        3 => "three",
        4 => "four",
        5 => "five",
        6 => "six",
        7 => "seven",
        8 => "eight",
        9 => "nine",
        10 => "ten",
    ];
    
    public function verse(int $number): string
    {
        $numeral = mb_ucfirst(self::NUMERALS[$number]);
        $next_numeral = self::NUMERALS[$number - 1];
        $bottle_plural = $number === 1 ? 'bottle' : 'bottles';
        $next_bottle_plural = $number === 2 ? 'bottle' : 'bottles';
        return <<<TEXT
            {$numeral} green {$bottle_plural} hanging on the wall,
            {$numeral} green {$bottle_plural} hanging on the wall,
            And if one green bottle should accidentally fall,
            There'll be {$next_numeral} green {$next_bottle_plural} hanging on the wall.
            TEXT;        
    }

    public function verses(int $start, int $size): string
    {
        $verses = [];
        for ($i = $start; $i > $start - $size; $i--) {
            $verses[] = $this->verse($i);
        }
        return implode("\n\n", $verses);
    }

    public function lyrics(): string
    {
        return $this->verses(10, 10);
    }
}

File: solutions/r/bottle-song/1/bottle-song.R
recite <- function(start_bottles, take_down) {
  # Map numbers to their English title-case equivalents
  num_words <- c("no", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten")
  
  # Helper function to generate a single verse
  get_verse <- function(n) {
    # Capitalize the first letter of the number word for the start of lines
    current_word <- num_words[n + 1]
    current_cap <- paste0(toupper(substring(current_word, 1, 1)), substring(current_word, 2))
    
    next_word <- num_words[n]
    
    # Singular vs. plural handling for "bottle" vs "bottles"
    current_bottle_str <- if (n == 1) "bottle" else "bottles"
    next_bottle_str <- if (n - 1 == 1) "bottle" else "bottles"
    
    line1_2 <- paste0(current_cap, " green ", current_bottle_str, " hanging on the wall,")
    line3   <- "And if one green bottle should accidentally fall,"
    line4   <- paste0("There'll be ", next_word, " green ", next_bottle_str, " hanging on the wall.")
    
    return(c(line1_2, line1_2, line3, line4))
  }
  
  # Generate all required verses in descending order
  verses <- list()
  for (i in seq_len(take_down)) {
    current_bottles <- start_bottles - i + 1
    verses[[i]] <- get_verse(current_bottles)
  }
  
  # Collapse each verse with newlines, and join multiple verses with a blank line
  verse_strings <- sapply(verses, function(v) paste(v, collapse = "\n"))
  full_song <- paste(verse_strings, collapse = "\n\n")
  
  return(full_song)
}